### DISCUSSION TITLE
Question: Use at build time?

### DISCUSSION THREAD

**Discussion by M-Schiller** [mganss/XmlSchemaClassGenerator#519] [Q&A] [answered]
Thank you for creating this amazing tool and apologies for posting a question here, but I could not find a guide to or usage of what I would like to have in a project I am developing.

Currently, I use a version of `XmlSchemaClassGenerator.Console` that I need to check into a repository (under `tools\..`) and use in my project as

```xml
  <Target Name="Create classes from schema" BeforeTargets="PreBuildEvent" AfterTargets="Clean">
      <Message Importance="high" Text="Running batch script..." />
      <Exec Command=".\tools\xscgen\XmlSchemaClassGenerator.Console.exe .\data\MySchema.xsd" />
  </Target>
```

What I am trying to achieve is a way of handling this via the NuGet package and without having to install the dotnet tool (if possible) so that I can also use the functionality in a build pipeline that does not have any tools installed.

**Comment by mganss** [comment, marked as answer]
I think the easiest way is to use xscgen as a [local tool](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-local-tool).

Another option of course is to use the XmlSchemaClassGenerator library in a custom console app that can be part of your solution. This may sound harder than it is, should be only a few lines.

In #281 @shuebner started an effort to create a source generator which would probably be ideal for your use case, but unfortunately it has stalled.

**Reply by M-Schiller** [reply]
Thank you for your response and for replying as quickly as you did.

What functionality is missing from the mentioned proof of concept to become a MVP?

**Reply by shuebner** [reply]
@M-Schiller

I trust you already had a look at my repo for the generator here: https://github.com/shuebner/XsdToSource
IIRC what is missing is what I wrote in the PR:

Supporting multiple schema files.

Adding generation options. All the options available for the xscgen tool should be available via some mechanism that is readily consumable by the source generator, e. g. through custom metadata on the `AdditionalFiles` item, like the PoC option `XsdToSource_RootNamespace` already is (see [Sample/Sample.csproj](https://github.com/shuebner/XsdToSource/blob/49164ad039b91b2e7ae4f03ce23c0609f6da21dc/Sample/Sample.csproj#L13))

I would estimate that this could be turned into an MVP in a week.
But I do not have time for that.

So if you are not willing to invest that time, then the xscgen tool is the next best option, especially since using dotnet tools is well-supported nowadays:

1. create a tool manifest and make sure that `dotnet tool restore` is part of your build (both local and CI).
2. add an MSBuild target to your project file that is executed before the actual build and uses the tool to create the generated cs files, like so (untested, there may be mistakes):
```csproj
<Project>
  <!-- here is all the other stuff -->

  <Target Name="__GenerateSourceFromXsd" BeforeTargets="BeforeBuild">
    <!-- remove previously generated code -->
    <ItemGroup>
      <Compile Remove="$(ProjectDir)generated/*.cs" />
    </ItemGroup>
    <Delete Files="$(ProjectDir)generated/*" />

    <!-- generate the cs files from the schema. this assumes the tool is available at this point, e. g. by having done "dotnet tools restore" with the appropriate tool manifest -->
    <Exec WorkingDirectory="$(ProjectDir)" Command="dotnet xscgen <alltheoptions>" />

    <!-- add generated files to the compilation -->
    <ItemGroup>
      <Compile Include="$(ProjectDir)generated/*.cs" />
    </ItemGroup>
  </Target>
</Project>
```

**Comment by shuebner** [comment]
@mganss rubbing it in, I see ;-). Yes, I have little time and so many other things to do, sorry. Unfortunately noone has picked it up where I left off.

But thanks for remembering me at all after all this time.
